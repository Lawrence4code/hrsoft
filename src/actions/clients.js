import uuid from 'uuid';
import database from '../firebase/firebase';
import { Children } from 'react';
// action generator for adding client
// passed with default value if no value passed for the client
export const addClient = client => ({
  type: 'ADD_CLIENT',
  client
});

export const startAddClient = (clientData = {}) => {
  console.log('startAddClient triggered');
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      firstName = '',
      lastName = '',
      gender = '',
      phone = '',
      email = '',
      address = '',
      status = '',
      field = '',
      note = '',
      lastCommuniation = '',
      memberType = '',
      createdAt = 0
    } = clientData;

    const client = {
      title,
      firstName,
      lastName,
      gender,
      phone,
      email,
      address,
      status,
      field,
      note,
      lastCommuniation,
      memberType,
      createdAt
    };

    // console.log(client);

    return database
      .ref(`users/${uid}/clients`)
      .push(client)
      .then(ref => {
        // console.log(ref.key);
        dispatch(
          addClient({
            id: ref.key,
            ...client
          })
        );
      });
  };
};

// action generating for removing client
export const removeClient = ({ id } = {}) => {
  return {
    type: 'REMOVE_CLIENT',
    id
  };
};

export const startRemoveClient = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients/${id}`)
      .remove()
      .then(() => {
        dispatch(removeClient({ id }));
      });
  };
};

// action generating for modifying the client details
export const editClient = (id, updates) => {
  return {
    type: 'EDIT_CLIENT',
    id,
    updates
  };
};

export const startEditClient = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editClient(id, updates));
      });
  };
};

// set clients

export const setClients = clients => {
  return {
    type: 'SET_CLIENTS',
    clients
  };
};

// export const startSetClients

export const startSetClients = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/clients`)

      .once('value')
      .then(snapshot => {
        const clients = [];
        snapshot.forEach(childSnapshot => {
          clients.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setClients(clients));
      });
  };
};
