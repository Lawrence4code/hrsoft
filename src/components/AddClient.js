import React from 'react';
import { Link } from 'react-router-dom';

const AddClient = () => (
  <div>
    <h2>AddClientPage</h2>
    <Link to="/"> Back to Dashboard </Link>
  </div>
);

export default AddClient;
