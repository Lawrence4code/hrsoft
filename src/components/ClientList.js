import React from 'react';
import { connect } from 'react-redux';
import ClientListItem from './ClientListItem';
import getVisibleClients from '../selectors/clients';

export const ClientList = props => (
  <div>
    {props.clients.length === 0 ? (
      <p> No Clients </p>
    ) : (
      props.clients.map(client => {
        return <ClientListItem key={client.id} {...client} />;
      })
    )}
    {}
  </div>
);

const mapStateToProps = state => {
  // console.log(state);
  return {
    clients: getVisibleClients(state.clients, state.filters)
  };
};

export default connect(mapStateToProps)(ClientList);
