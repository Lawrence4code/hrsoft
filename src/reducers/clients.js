// default values for the cient reducer
const clientReducerDefaultState = [];

const clientsReducer = (state = clientReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_CLIENT':
      return [...state, action.client];
    case 'EDIT_CLIENT':
      return state.map(client => {
        if (client.id === action.id) {
          return {
            ...client,
            ...action.updates
          };
        } else {
          return client;
        }
      });
    case 'REMOVE_CLIENT':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};

export default clientsReducer;
