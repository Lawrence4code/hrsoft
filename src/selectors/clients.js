// get visible clients
import moment from 'moment';

const getVisibleClients = (clients, { name, sortBy, startDate, endDate }) => {
  // console.log(clients);
  // console.log(startDate, endDate);
  return clients
    .filter(client => {
      const createdAtMoment = moment(client.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, 'day')
        : true;

      const firstNameMatch = client.firstName
        .toLowerCase()
        .includes(name.toLowerCase());
      const lastNameMatch = client.lastName
        .toLowerCase()
        .includes(name.toLowerCase());

      return (
        startDateMatch && endDateMatch && (firstNameMatch || lastNameMatch)
      );
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'alphabetically') {
        console.log('alpha sort');
        return -1;
      }
    });
};

export default getVisibleClients;
