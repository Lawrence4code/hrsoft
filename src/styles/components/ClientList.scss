.clientListHeading {
  display: grid;
  grid-template-columns: 1.55fr 1.85fr 1.25fr 80px 1fr 100px 100px 100px 100px;
  text-align: center;
  margin: 0 auto 1rem auto;
  width: 90%;
  background-color: #a9a9d8;
  align-items: center;
  color: white;
  border-radius: 3px;
  font-weight: bold;

  & p {
    text-transform: uppercase;
    font-size: 12px;
    padding: 0.8rem 0;
  }
}
