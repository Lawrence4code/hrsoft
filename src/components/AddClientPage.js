import React from 'react';
import ClientForm from './ClientForm';
import { connect } from 'react-redux';
import { startAddClient } from '../actions/clients';

export class AddClientPage extends React.Component {
  onSubmit = client => {
    this.props.startAddClient(client);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="addClientPage">
        <div className="addClientPage__heading">
          <div className="addClientPage__content">
            <h1> Add New Client </h1>
          </div>
        </div>

        <ClientForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddClient: client => dispatch(startAddClient(client))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddClientPage);
