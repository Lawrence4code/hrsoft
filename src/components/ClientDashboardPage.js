import React from 'react';
import { Link } from 'react-router-dom';
import ClientList from './ClientList';
import ClientListFilters from './ClientListFilters';
import ClientSummary from './ClientSummary';

const ClientDashboardPage = () => (
  <div>
    <ClientSummary />
    <ClientListFilters />
    <ClientList />
  </div>
);

export default ClientDashboardPage;
