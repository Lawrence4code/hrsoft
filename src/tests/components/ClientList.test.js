import React from 'react';
import { shallow } from 'enzyme';
import { ClientList } from '../../components/ClientList';
import clients from '../fixtures/clients';
import { wrap } from 'module';

test('should render ClientList with client details', () => {
  const wrapper = shallow(<ClientList clients={clients} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ClientList with empty message', () => {
  const wrapper = shallow(<ClientList clients={[]} />);
  expect(wrapper).toMatchSnapshot();
});
