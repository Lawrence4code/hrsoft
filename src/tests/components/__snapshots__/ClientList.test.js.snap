// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ClientList with client details 1`] = `
<div>
  <Connect(Component)
    address="FakeOne High Street"
    createdAt={0}
    email="John@doe.com"
    field="Designer"
    firstName="John"
    gender="M"
    id={1}
    key="1"
    lastCommuniation=""
    lastName="Doe"
    memberType="Gold"
    note="Note1"
    phone="987654321"
    status="Employed"
    title="Mr"
  />
  <Connect(Component)
    address="FakeOne High Street"
    createdAt={-345600000}
    email="Jane@doe.com"
    field="n/a"
    firstName="Jane"
    gender="F"
    id={2}
    key="2"
    lastCommuniation=""
    lastName="Doe"
    memberType="Silver"
    note="Not needed"
    phone="123456789"
    status="Housewife"
    title="Mrs"
  />
  <Connect(Component)
    address="FakeTwo High Street"
    createdAt={345600000}
    email="Peter@doe.com"
    field="Teacher"
    firstName="Peter"
    gender="M"
    id={3}
    key="3"
    lastCommuniation=""
    lastName="Doe"
    memberType="Gold"
    note="Note3"
    phone="9999999999"
    status="Student"
    title="Mr"
  />
</div>
`;

exports[`should render ClientList with empty message 1`] = `
<div>
  <p>
     No Clients 
  </p>
</div>
`;
